### ISSUE TITLE
[RFC] Improvements to the catalog backend

### ISSUE THREAD

**Issue by GoWind** [backstage/backstage#4961]
**Status:** Open for comments

## Need

Performance of the `catalog-backend` can be improved significantly.

This RFC was created out of a need that was observed when deploying Backstage @ Zalando.

The catalog backend's database solution is simple and functions well, but we noticed significant slowdowns and poor ability to scale even with a modest amount of entities. 

We roughly have 20k entries in the `entities` table. 

In the current design of the frontend and the backend plugins, performance is found wanting.

For example, when querying the catalog table for data, the `catalog` plugin , by default attempts to query all `Component` entities from the table, resulting in a response of about 5k entries that takes 2 secs on the average per request. 

These requests also , possibly due to the use of an ORM, overload the instance (with about 4GB of memory), prompting restarts.  

From our testing, we were unable to handle loads > 8 Requests per second. 

The frontend's default queries can be customized , as part of other PRs, but this RFC is about improving the design of the `catalog-backend`'s tables.

### Searching and querying
Searching and querying for entities are very dissimilar actions and should ideally be handled separately. Given that most entities have a section with specific fields and values (`apiVersion`, `kind`, `metadata.name`, `spec.type`) and a schema-less section (`spec`), we should ideally split the functionality currently satisfied by `entities_search`.  

Searching for entities based on arbitrary path expressions, filters and ranges should be handled ideally by a GIN software/index such as elastic search. 

Querying, for specific fields and values can be done much faster and will solve a majority of use-cases.

## Proposal

This proposal attempts to improve the layout of the backend's table without too much disruption.
1. Add indexes on the `entities` table for specified fields : `apiVersion`, `kind`, `metadata.name` (in conjunction with `namespace` and `kind`) and `spec.type`. 
2. Re-think the idea of using a `text` field to store the contents of an entity. A `JSON` or a `JSONB` storage type comes to mind. 
`JSONB` has the advantage that it supports indexing nested fields, path queries and more compact storage, but has some caveats related to character encoding, that must be kept-in [mind](https://www.postgresql.org/docs/9.4/datatype-json.html)

**Note**: IMHO, I still think searching for documents based on ranges of values or matching the closes possible document should still be done by a specialized tool such as `Elasticsearch`, but a `JSONB` structure will perhaps make a subset of such queries easier. 

3. Allow or discuss the possibility and ideas of providing a raw `SQL` interface. While ORMs are easy to prototype and start with, they might not produce optimal queries and often become bottlenecks. It might and should be possible for implementations to be able to run custom queries instead of using `knex`.

**Note**: no. 3 is still a very vague idea in my head and could use some brainstorming from the community. 

As we ramp up on the usage of  the Catalog and Backstage, we figured out that this might be a relevant issue for others in the community as well and therefore begs a public discussion.

**Comment by Fox32** [comment]
One additional improvement might be #4007 especially if you have a lot of API entities in the catalog.

**Comment by GoWind** [comment]
We still haven't started holding APIs in the catalog, but it is highly likely we will run into the same problem, highlighted in #4007. 

OTOH, We can solve a lot of problems mentioned in 4007, by bumping up the version of the `API` Entity Kind that supports linking to an external URL where the actual API definition is stored.

As to blob storage in the catalog, I am rather cautious and would tend to avoid it, not because of performance issues, but any `bag` like API would tend to be abused to store all sorts of irrelevant data in the catalog and any future migrations / technial debt resolution will become tricky.  But this is my opinion/heuristics, so please feel free to correct my assumptions.

**Comment by freben** [comment]
Partly addressed by #4993

**Comment by freben** [comment]
> Allow or discuss the possibility and ideas of providing a raw SQL interface. While ORMs are easy to prototype and start with, they might not produce optimal queries and often become bottlenecks. It might and should be possible for implementations to be able to run custom queries instead of using knex.

Knex does allow falling through to entirely raw queries, using `tx.raw('SQL', [variables])`.

**Comment by Rugvip** [comment]
@GoWind @freben Knex is not an ORM, it's just a layer that abstracts away differences in SQL dialects and provides some utilities like migrations. We're essentially already writing raw SQL, just through a JavaScript interface.

Regarding JSONB and indexing the entity contents I somewhat agree, it could help to avoid the extra `entities_search` table, and for SQLite we probably just fall back to filtering in the backend instead. Would be a bit of a rework though.

We will be bringing in Elasticsearch to support search, but it will be done in a separate search system not directly tied to the catalog. The catalog APIs themselves will largely be kept focused at more manual browsing and filtering, at least that's the direction right now.

**Comment by GoWind** [comment]
> We will be bringing in Elasticsearch to support search, but it will be done in a separate search system not directly tied to the catalog. The catalog APIs themselves will largely be kept focused at more manual browsing and filtering, at least that's the direction right now.

:+1:  to this . I would agree on keeping elasticsearch separate for now. 

What I meant by a `raw` interface was a way to install custom handlers  for API endpoints on the catalog-backend. 

The motivation for this requirement stems from the fact that we have to do a lot of temporary mappings, APIs and migrations to eventually move the content in existing systems to the catalog-backend. 

These usecases are currently limited by the querying/filter interface in the catalog API, but most importantly, doing all these mappings in the same runtime as the `catalog` backend will provide us simpler mechanisms to eventually migrate to the catalog database. 

Another improvement we wanted to do was allow the `refresh` of locations in a separate thread. 
Its not very compute intensive, but doing refreshes in the same thread handling the application requests makes the memory and CPU consumption patterns of the instances we are running the service on, very unpredictable and bursty.

Our application request serving needs have more or less stable patterns and we can provision predictable hardware capacity, whereas the locations refresh is bursty and offers minimal mechanisms for spreading the load across the day. 

My suggestions would be as follows
1. Investigate and allow the `refresh` to occur in a separate thread, via configuration. This will prevent the main thread from being blocked serving what is basically a background, cron job. 
2. allow refreshing location by `kind`, `spec.type` or another key . This can be used to schedule different refreshes during different periods of the day, to distribute the load more evenly. 

@Fox32  : Do you see any concerns/additions here that might be of use to you as well ? 


I would volunteer myself for any assistance (or PRs) for this topic. We would soon start needing more features from the catalog-backend, so I guess I can start with finding  a way to add a new `entities` schema, without disrupting the existing `entities`. 
This can be done in 2 stage, 
1. Add indexes for `kind`, `spec.type` and `apiVersion` on the existing`entities` table
2. Create a new JSON based `entities` schema. 
3.Provide support in backstage for  migrating entities from existing `entites` to the new `JSON` entities.
2. Drop the `entities` schema.

Obviously a lot of these changes are going to be breaking changes, so I will have to synchronize with the backstage-core on what you plans you have for the catalog-backend. 

Please do let me know what you think of the `raw` API endpoint idea . I admit its not very descriptive, but perhaps someone lese has an usecase for this.

**Comment by Rugvip** [comment]
@GoWind We're currently knee-deep in reworking the catalog backend. Here's the rought plan so far:

- Refresh locations and entities individually rather than all at once. To achieve this we essentially add a processing queue in the database that's shared across instances, allowing the catalog processing to scale horizontally.
- Lazily processing discovered entities and locations rather than the eager way it is done now. This means that when we discover an entity we put it on the queue for later processing rather than processing it straight away. It still get picked from the queue quickly though.
- Allow entities to store state during processing that is provided back when they are processed again, allowing for caching of resources and avoiding unnecessary updates.
- Seed the processing with a location provider interface, allowing for a more deliberate way of feeding the processing loop rather than the somewhat awkward bootstrap location.
- Split out the registered locations store into a more well separated controller and store, and plugging it in to the processing via the new provider interface.
- Make location entities first class entities that are stored and read back from the catalog, right now they're just used for processing then thrown away. This will make them part of the catalog graph and will allow us to store error information on them etc.
- If possible, get rid of the relations table where entity relations are stiched together read time, and more this to processing instead. This should improve read performance.

Regarding processing in a separate thread, we're discussing using worker threads, but given that it's a completely separate V8 context it's something that needs to be set up in the backend and likely needs build tooling support as well.

I'd also like for us to provide a nicer interface at the controller/service level, just after the router, which you can then decorate to your heart's content :grin:

**Comment by GoWind** [comment]
Sounds exciting. I can hardly wait ! 

How do you figure out fitting in the indexes and the JSON data types feature into this ? 
Do you think this can be done in parallel, or weaved into your plan ? 
We can implement some of the features and then bring it to backstage (or vice versa, solve it in backstage). 

While we were discussing internally, yet another issue came up:  **aliases**.

Currently, the tuple (`namespace`, `kind`,`metadata.name`) is unique and has a corresponding UUID. 

In large organizations, it is not uncommon to have many entities with a unique ID , but with a common, Human readable name. For example there could be multiple `inventory` teams where `inventory` is dependent on the context. The bigger the organization, the more likely such cases exist.

This is true in our case.  Multiple teams have aliases that are referenced in certain sub-systems in different ways.

We also have some issues mapping APIs into the catalog.

We have roughly 2k+ APIs and they are identified by an API ID, that is not human readable. Each API has multiple revisions (rev1, 1.1, etc). Currently any number of revisions might be deployed across the entire tech infrastructure.

This does not fit in well with the Entity structure of having a unique name for each entity. To solve this problem, the solution would be most likely a hack (set  name as `${id}+${revision}`, but then it no longer becomes human readable. 

To solve this problem, we thought of adding fields to the `entity` such as `spec.searchableNames` or `spec.aliases` : `[]string`, but I am not sure how to add an index on this field. I guess the `CatalogClient` can take this and filter for entities that have this field (can it) , but this is slow, due to the design of the `entities_search` table. 

**Question**: Do you think a first class Aliasing solution (maybe an `Alias` entity or a `relation` or both ? ) It could also be as simple as an indexable `alternateNames` or `aliases` field in the `spec`, that can be used to look up an entity UUID, based on a given `alias`. 

I proposed Aliasing as a solution on Backstage, because I believe it might be a common enough problem to solve via Backstage and not require implementations to come up with hacked  together implementations of their own.

**Comment by tkrop** [comment]
To second @GoWind´s request. Our problem is mainly the presentation perspective. We have no problem generating sufficient unique `name`s, but these are effectively no `name`s any longer - they are `id`s. While we miss the ability to represent entities by a human readable (and not naturally unique) `alias`, i.e. `name`.

The main point here is, that `id`s and `name`s are to properties that starting at a certain number cannot any longer kept in sync. Thus, in our opinion, it makes sense to split this functionally and provide for both separate support in Backstage. As a company we have the additional problem, that we often have multiple `name`s (aliases) for the same entity (especially in the case of teams), since things move fast and we need support the old naming until everything is updated.

**Comment by freben** [comment]
I've been considering adding `metadata.title`, which the Template kind already makes use of. It would be optional and used exclusively for presentation purposes, and `metadata.name` would be used as the fallback.

This is a loose idea, not a declaration of intent :) does it match what your needs are?

**Comment by freben** [comment]
(Mostly directed at @tkrop)

**Comment by GoWind** [comment]
@freben  Should mostly be sufficient. 

Rather than (Or, in addition to) adding support here, I think it would be better to allow customization of the Catalog Frontend plugin to display an arbitrary `name` on the initial catalog table and in subsequent Entity cards.

Currently in the home view of the catalog plugin, it is the `metadata.name` of the component that is shown.  

Given that the `spec` of an Entity is arbitrary, rather than extend the model with yet another required/optional/well-known field, I think it would be a simpler solution to customize the UI to display an arbitary key from the Entity's description (which is also exactly what our original requirement is). 

We could treat the `spec` as a map of properties to values and can use any value in the `spec` without distrubing the backstage Entity model.

**Comment by GoWind** [comment]
As for slow responses, I did some profiling today and figured out what might be one of the causes 

In the function `toEntityResponses` (https://github.com/backstage/backstage/blob/dddf2bc5d63df24b7af499c575b5f86656778271/plugins/catalog-backend/src/database/CommonDatabase.ts#L494),  the relations for each entity is fetched using a call to the `getRelationsByFullname` function 
https://github.com/backstage/backstage/blob/dddf2bc5d63df24b7af499c575b5f86656778271/plugins/catalog-backend/src/database/CommonDatabase.ts#L500

The `getRelationsByFullname` function, reads the relations for the entities by using an `IN` clause on the `fullName`).

However, each entity in the entity response (`DbEntitiesRow`) already has an `id` field in memory.
The relations also have an `originating_entity_id` field. 
Instead of searching using `IN` on a text match, this can be replaced by a `JOIN ON row.id = relations.originating_entity_id`, which should be IMHO, much much faster, as they are done on the primary key fields which should be indexed. 

I have not profiled this `join`, but from experience, if this can be done, it should be much faster. Would you like a PR to improve this ? I am willing to debug this and raise one if we get much  improved performance.

**Comment by freben** [comment]
@GoWind If you find it to be both faster on pg, and at least of similar speed on sqlite, then sure. See this comment https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/CommonDatabase.ts#L222 that is the result of me being down that path before and having to reverse the decision.

Since then, the indices have changed slightly, so it may be out of date

**Comment by freben** [comment]
I should also mention, that as part of the catalog rewrite, we may choose to materialise the relations directly into the resulting entities. If we do that, the join will go away entirely.

**Comment by GoWind** [comment]
Noted, I will inspect the sql statements generated for these filters/queries and see if I can find an improvment.

**Comment by Rugvip** [comment]
> However, each entity in the entity response (`DbEntitiesRow`) already has an `id` field in memory. The relations also have an `originating_entity_id` field. Instead of searching using `IN` on a text match, this can be replaced by a `JOIN ON row.id = relations.originating_entity_id`, which should be IMHO, much much faster, as they are done on the primary key fields which should be indexed.

We actually need to look up by the name, we're querying for matches for `source_full_name`. `originating_entity_id` is something different and is used to delete and overwrite relations, but it's not used in the stitching together of the final entity.

As freben mentioned we'll be trying to move that step away from read time anyway :grin:

**Comment by GoWind** [comment]
Interesting results .  `where in` and `join` on `id` have almost the same cost. Did some analysis using `explain query`

```
backstage_plugin_catalog=> explain select r.* from entities_relations r join entities e on e.full_name = r.source_full_name;
                                    QUERY PLAN                                    
----------------------------------------------------------------------------------
 Hash Join  (cost=3636.66..5054.97 rows=48894 width=77)
   Hash Cond: ((r.source_full_name)::text = e.full_name)
   ->  Seq Scan on entities_relations r  (cost=0.00..1289.94 rows=48894 width=77)
   ->  Hash  (cost=3367.96..3367.96 rows=21496 width=26)
         ->  Seq Scan on entities e  (cost=0.00..3367.96 rows=21496 width=26)
(5 rows)


backstage_plugin_catalog=> explain select r.* from entities_relations r join entities e on e.full_name = r.source_full_name;
                                    QUERY PLAN                                    
----------------------------------------------------------------------------------
 Hash Join  (cost=3636.66..5054.97 rows=48894 width=77)
   Hash Cond: ((r.source_full_name)::text = e.full_name)
   ->  Seq Scan on entities_relations r  (cost=0.00..1289.94 rows=48894 width=77)
   ->  Hash  (cost=3367.96..3367.96 rows=21496 width=26)
         ->  Seq Scan on entities e  (cost=0.00..3367.96 rows=21496 width=26)
(5 rows)
```
We have roughly 48k entries in relations and 21k entries in entities.

**Comment by GoWind** [comment]
But there does seem to be a sizeable estimated cost difference when filter for the initial set of entities of a particular `type`.

Query 1 is I  think how the current filter filters for entities (but in memory).
```
backstage_plugin_catalog=> explain select * from entities where id in (select entity_id from entities_search where key like 'spec.type%' and value like 'service');
                                             QUERY PLAN                                              
-----------------------------------------------------------------------------------------------------
 Nested Loop  (cost=7301.73..8158.65 rows=126 width=1021)
   ->  HashAggregate  (cost=7301.44..7302.70 rows=126 width=16)
         Group Key: entities_search.entity_id
         ->  Bitmap Heap Scan on entities_search  (cost=163.78..7301.13 rows=126 width=16)
               Filter: (((key)::text ~~ 'spec.type%'::text) AND (value ~~ 'service'::text))
               ->  Bitmap Index Scan on entities_search_value  (cost=0.00..163.75 rows=4176 width=0)
                     Index Cond: (value = 'service'::text)
   ->  Index Scan using entities_pkey on entities  (cost=0.29..6.84 rows=1 width=1021)
         Index Cond: (id = entities_search.entity_id)
(9 rows)
```
A join directly instead of a select and select, seems to have a much smaller startup cost (an order of magnitude)
```
backstage_plugin_catalog=> explain select e.* from entities e join  entities_search s on e.id = s.entity_id where s.key like 'spec.type%' and s.value like 'service';
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Nested Loop  (cost=164.06..8163.56 rows=126 width=1021)
   ->  Bitmap Heap Scan on entities_search s  (cost=163.78..7301.13 rows=126 width=16)
         Filter: (((key)::text ~~ 'spec.type%'::text) AND (value ~~ 'service'::text))
         ->  Bitmap Index Scan on entities_search_value  (cost=0.00..163.75 rows=4176 width=0)
               Index Cond: (value = 'service'::text)
   ->  Index Scan using entities_pkey on entities e  (cost=0.29..6.84 rows=1 width=1021)
         Index Cond: (id = s.entity_id)
(7 rows)
```
One of the my motivations for a `raw` interface was that I use it to compare and optimize my queries against that of `knex`.

**Comment by Crevil** [comment]
Just a quick note on the `explain` results. Be aware of caching. 😃 I'm not sure if this is the case here, but I have personally drawn wrong conclusions based on cached results before 🙈

**Comment by freben** [comment]
Hm are those exactly what knex issued? The `s.key like 'spec.type%'` is unexpected since we do equality testing with `this.andWhere({ key: key.toLowerCase() });` - should not

**Comment by GoWind** [comment]
@freben : I should have qualified my statements better

> Query 1 is I think how the current filter filters for entities (but in memory).

Query 1 is NOT the query run by knex, but how I assumed it would generate queries.

Nevertheless, I think from the snippet you [linked](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/database/CommonDatabase.ts#L222), to me it sounds like the actual filtering is done , perhaps in application memory (because you pass a key js `fn` ?) and not exactly in the database. 

In any case, I will have to find ways to get the actual SQL statements generated, logged in the console for debugging (perhaps this is a nice improvement to the backend as well ? )

**Comment by GoWind** [comment]
@Crevil  : Haha, Noted ! Thanks. I will check for caching of query plans/results when doing `explain`s

**Comment by tkrop** [comment]
@freben Sorry for coming back that late: I love the idea. Adding a `metadata.title` that has no unique constraint on it like `metadata.name` and making use of it in catalog presentation as primary field would definitely solve most of our problems. The `metadata.name` should only provide a fallback, if no `metadata.title` is available.

However, this would also change how people navigate: keywords in `metadata.title` will likely become the default terms used for searching entities. Anyhow, how can we help on delivering this. Should we open an issue or contribute with a pull request?
